modules/userman/oauth.go

```diff
@@ -0,0 +1,281 @@
+		ClientID:     c.ID,
+		ClientSecret: c.Secret,
+		RedirectURL:  c.Host + "/v1/oauth/google-callback",
+		Endpoint:     googleOAuth2.Endpoint,
+		Scopes:       []string{"profile", "email"},
+	}
+
+	// state param cookies require HTTPS by default; disable for localhost development
+	//stateConfig := gologin.DebugOnlyCookieConfig
+	stateConfig := gologin.DefaultCookieConfig
+	router.Handle("/v1/api/{project}/auth/{dbType}/oauth/google", google.StateHandler(stateConfig, google.LoginHandler(oauth2Config, nil)))
+	router.Handle("/v1/api/{project}/auth/{dbType}/oauth/google-callback", google.StateHandler(stateConfig, google.CallbackHandler(oauth2Config, issueGoogleSession(m), nil)))
+}
+
+func issueGoogleSession(m *Module) http.Handler {
+	fn := func(w http.ResponseWriter, r *http.Request) {
+		// Get the path parameters
+		vars := mux.Vars(r)
+		project := vars["project"]
+		dbType := vars["dbType"]
+
+		ctx := r.Context()
+		googleUser, err := google.UserFromContext(ctx)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		token, err := m.addUser(ctx, project, dbType, googleUser.Email, googleUser.Name)
+		if err != nil {
+			log.Println("Err: ", err)
+			w.WriteHeader(http.StatusInternalServerError)
+			json.NewEncoder(w).Encode(map[string]string{"error": err.Error()})
+			return
+		}
+
+		// Get the OAuth config
+		c, _ := m.getOAuth("google")
+
+		http.Redirect(w, r, c.RedirectURL+"?token="+token, http.StatusSeeOther)
+	}
+	return http.HandlerFunc(fn)
+}
+
+func (m *Module) newFacebookOAuth(router *mux.Router) {
+	// Don't register routes if module is disabled
+	if enabled := m.isActive("fb"); !enabled {
+		return
+	}
+
+	// Get the OAuth config
+	c, _ := m.getOAuth("fb")
+
+	oauth2Config := &oauth2.Config{
+		ClientID:     c.ID,
+		ClientSecret: c.Secret,
+		RedirectURL:  c.Host + "/v1/oauth/fb-callback",
+		Endpoint:     facebookOAuth2.Endpoint,
+		Scopes:       []string{"email"},
+	}
+
+	// state param cookies require HTTPS by default; disable for localhost development
+	//stateConfig := gologin.DebugOnlyCookieConfig
+	stateConfig := gologin.DefaultCookieConfig
+	router.Handle("/v1/api/{project}/auth/{dbType}/oauth/fb", facebook.StateHandler(stateConfig, facebook.LoginHandler(oauth2Config, nil)))
+	router.Handle("/v1/api/{project}/auth/{dbType}/oauth/fb-callback", facebook.StateHandler(stateConfig, facebook.CallbackHandler(oauth2Config, issueFacebookSession(m), nil)))
+}
+
+func issueFacebookSession(m *Module) http.Handler {
+	fn := func(w http.ResponseWriter, r *http.Request) {
+		// Get the path parameters
+		vars := mux.Vars(r)
+		project := vars["project"]
+		dbType := vars["dbType"]
+
+		ctx := r.Context()
+		facebookUser, err := facebook.UserFromContext(ctx)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		token, err := m.addUser(ctx, project, dbType, facebookUser.Email, facebookUser.Name)
+		if err != nil {
```

Comment:
Why are we using different techniques for handling error?

Reply:
Just